title: "1. Human Subtype-IVYGAP Validation"
author: "Daan Kloosterman"
date: "8/12/2021"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Validating the glioblastoma cellular subtype and pseudolocation classification function

# Loading packages/libraries required for the code

```{r}
library(Seurat)
library(nichenetr)
library(RColorBrewer)
library(openxlsx)
library(ggpubr)
library(gridExtra)
library(gprofiler2)
```


# Setting working directory and loading all the pre-processed data files (only works when you press "Run Current Chunk")

```{r}
# Set working directory to folder "pre-processed human scrna-seq" 
work_dir <- "~/surfdrive/Code Availability/Kloosterman_et_al/Kloosterman et al. 2022 - data availibility/GEO Data - processed/scrna-seq"
setwd(work_dir)

load("gbm-complete-tumor_ink4a.Rda")
load("gbm-complete-tumor_p53.Rda")

features <- SelectIntegrationFeatures(object.list = c(gbm.tumor_ink4a, gbm.tumor_p53), nfeatures = 3000)
gbm.anchors <- FindIntegrationAnchors(object.list = c(gbm.tumor_ink4a, gbm.tumor_p53),  anchor.features = features)
rm(gbm.list)

gbm.tumor <- IntegrateData(anchorset = gbm.anchors)
rm(gbm.anchors)

gbm.tumor <- NormalizeData(gbm.tumor)
gbm.tumor <- FindVariableFeatures(gbm.tumor, selection.method = "vst", nfeatures = 2000)
gbm.tumor <- ScaleData(gbm.tumor, verbose = FALSE)
gbm.tumor <- FindVariableFeatures(gbm.tumor, verbose = FALSE)
gbm.tumor <- RunPCA(gbm.tumor, npcs = 5, verbose = FALSE)
gbm.tumor <- RunUMAP(gbm.tumor, reduction = "pca", dims = 1:5)

save(gbm.tumor,file="~/Desktop/gbm-complete-polished.Rda")
```


# Figure S1A-B and S1D

```{r}
## Plot tumor model/stage, subtype and location
DimPlot(gbm.tumor, group.by = "stage_model", pt.size = 1 ) + ggtitle("Background and stage") # Figure S1A
DimPlot(gbm.tumor, group.by = "Cell.Subtype", cols = c( "darkolivegreen3", "pink", "tomato3","darkgoldenrod1", "darkgoldenrod3", "royalblue3" ), pt.size = 1 ) + ggtitle("Cellular subtype") + NoLegend() # Figure S1B
DimPlot(gbm.tumor, group.by = "Location", cols = c("orange", "violetred1", "darkolivegreen3","lightblue", "grey"), pt.size = 1 ) + ggtitle("Pseudolocation") + NoLegend() # Figure S1D
```


# Find most differentially expressed genes across tumor subtype and niche and prepping heatmap

```{r}
DefaultAssay(gbm.tumor) <- "RNA"
gbm.tumor$spatialsubtype_grouped <- paste(gbm.tumor$Location ,gbm.tumor$Cell.Subtype)

# Finding classifiers for each sub population
gbm.tumor <- gbm.tumor[, gbm.tumor$spatialsubtype_grouped %in% c("LE MES1" ,"LE MES2" ,"LE OPC", "LE NPC1", "LE NPC2", "LE AC", "PAN MES1" ,"PAN OPC", "PAN NPC1", "PAN NPC2","PAN AC", "PAN MES2", "MVP MES1",  "MVP MES2","MVP OPC","MVP AC", "MVP NPC1", "MVP NPC2", "CT OPC", "CT MES1","CT MES2", "CT AC", "CT NPC1", "CT NPC2")]
Idents(gbm.tumor) <- gbm.tumor$spatialsubtype_grouped
pbmc.markers <- FindAllMarkers(gbm.tumor, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
pbmc.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)
top300 <- pbmc.markers %>% group_by(cluster) %>% top_n(n = 30, wt = avg_log2FC)
top300 <- as.character(top300$gene)
pbmc.markers <- as.character(pbmc.markers$gene)

df <- AverageExpression(object = gbm.tumor, group.by = "spatialsubtype_grouped", features = pbmc.markers)
df <- df$RNA[,1:24] # Only necessary when Undecided is not filtred from the dataset

subtype <- c("AC", "MES1","MES2", "NPC1","NPC2","OPC", "AC", "MES1","MES2", "NPC1","NPC2","OPC","AC", "MES1","MES2", "NPC1","NPC2","OPC","AC", "MES1","MES2", "NPC1","NPC2","OPC")
subtype <- as.data.frame(subtype)
location <- c("CT","CT","CT", "CT","CT", "CT" ,"LE","LE", "LE", "LE", "LE", "LE","MVP","MVP","MVP", "MVP", "MVP", "MVP","PAN", "PAN", "PAN", "PAN", "PAN", "PAN")
location <- as.data.frame(location)

df_row <- colnames(df)

rownames(subtype) <- df_row
colnames(subtype) <- c("Subtype")
rownames(location) <- df_row
colnames(location) <- c("Location")

anno_col <- cbind(subtype, location)
rownames(anno_col) <- df_row

## Get some nicer colours
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)

Location <- c("orange", "violetred1", "palegreen1","lightblue")
names(Location) <- c("CT", "LE", "MVP", "PAN")

Subtype <-c("darkolivegreen3", "pink", "tomato3","darkgoldenrod1", "darkgoldenrod3", "royalblue3" )
names(Subtype) <- c("AC", "MES1", "MES2", "NPC1", "NPC2", "OPC")

anno_colors <- list(Location = Location, Subtype = Subtype)
```


# Figure S1E: Heatmap of average gene expression derived from murine glioblastoma scRNA-seq dataset clustered by subtype and pseudolocation

```{r}
## Colors for heatmap
library(circlize)
library(ComplexHeatmap)

col_fun = colorRamp2(c(0, 0.5, 1), c("lightblue", "White", "orange"))
col_fun(seq(-3, 3))

df_norm <- t(apply(df, 1, function(x)(x-min(x))/(max(x)-min(x))))

split = data.frame(cutree(hclust(dist(df_norm)), k = 6))
out <- Heatmap(df_norm, col = col_fun, row_split = split, name = "Normalized Expression", show_row_names = F, row_title = " ", border = TRUE,
               left_annotation = rowAnnotation(foo = anno_block(gp = gpar(fill = 2:7),
        labels = c("1", "2", "3", "4", "5", "6"), 
        labels_gp = gpar(col = "black", fontsize = 10))),
        top_annotation = HeatmapAnnotation(
        Subtype = anno_col$Subtype,
        Niche = anno_col$Location,
        col = list(Subtype = c("AC" = "darkolivegreen3","MES1" = "pink","MES2" = "tomato3" ,"NPC1" = "darkgoldenrod1", "NPC2" = "darkgoldenrod3","OPC" = "royalblue3" ), Niche = c("CT" = "orange", "LE" = "violetred1", "MVP" = "palegreen1", "PAN" = "lightblue")),     
        border = TRUE,
        annotation_legend_param = list(
        Subtype = list(nrow = 1,  row_gap = unit(10, "mm")), Niche = list(nrow = 1,  row_gap = unit(10, "mm"))),
        show_annotation_name = F),
        
        show_column_names = F,
        cluster_columns = FALSE,
         heatmap_legend_param = list(
    at = c(0, 0.5, 1),
    labels = c("low", "zero", "high"),
    title = "Norm. Expression",
    legend_height = unit(4, "cm"), 
    direction = "horizontal"))  



draw(out,  merge_legend = F, heatmap_legend_side="bottom", annotation_legend_side="bottom", 
           legend_grouping = "original") # Save in 5x15


# 1 = 5, 2 = 6, 3 = 3, 4 = 4, 5 = 2, 6 = 1
```


# Average gene expression per group (there are 6 gene clusters)

```{r}
df_norm <- t(apply(df, 1, function(x)(x-min(x))/(max(x)-min(x))))

AC <- as.data.frame(df_norm[,c(1,7,13,19)])
colnames(AC) <- c("CT", "LE", "MVP", "PAN")
AC$gene <- rownames(AC)
AC$subtype <- c("AC")
AC$genecluster <- split$cutree.hclust.dist.df_norm....k...6.

MES1 <- as.data.frame(df_norm[,c(2,8,14,20)])
colnames(MES1) <- c("CT", "LE", "MVP", "PAN")
MES1$gene <- rownames(MES1)
MES1$subtype <- c("MES1")
MES1$genecluster <- split$cutree.hclust.dist.df_norm....k...6.

MES2 <- as.data.frame(df_norm[,c(3,9,15,21)])
colnames(MES2) <- c("CT", "LE", "MVP", "PAN")
MES2$gene <- rownames(MES2)
MES2$subtype <- c("MES2")
MES2$genecluster <- split$cutree.hclust.dist.df_norm....k...6.

NPC1 <- as.data.frame(df_norm[,c(4,10,16,22)])
colnames(NPC1) <- c("CT", "LE", "MVP", "PAN")
NPC1$gene <- rownames(NPC1)
NPC1$subtype <- c("NPC1")
NPC1$genecluster <- split$cutree.hclust.dist.df_norm....k...6.

NPC2 <- as.data.frame(df_norm[,c(5,11,17,23)])
colnames(NPC2) <- c("CT", "LE", "MVP", "PAN")
NPC2$gene <- rownames(NPC2)
NPC2$subtype <- c("NPC2")
NPC2$genecluster <- split$cutree.hclust.dist.df_norm....k...6.

OPC <- as.data.frame(df_norm[,c(6,12,17,24)])
colnames(OPC) <- c("CT", "LE", "MVP", "PAN")
OPC$gene <- rownames(OPC)
OPC$subtype <- c("OPC")
OPC$genecluster <- split$cutree.hclust.dist.df_norm....k...6.


combined <- rbind(AC, MES1,MES2, NPC1, NPC2, OPC)

CT <- as.data.frame(combined[,c(1, 5:7)])
CT$niche <- c("CT")
colnames(CT) <- c("Tumor")
LE <- as.data.frame(combined[,c(2, 5:7)])
LE$niche <- c("LE")
colnames(LE) <- c("Tumor")
MVP <- as.data.frame(combined[,c(3, 5:7)])
MVP$niche <- c("MVP")
colnames(MVP) <- c("Tumor")
PAN <- as.data.frame(combined[,c(4:7)])
PAN$niche <- c("PAN")
colnames(PAN) <- c("Tumor")

combined <- rbind(CT, LE, MVP, PAN)
colnames(combined) <- c("expression", "gene", "subtype", "genecluster", "niche")
  
# Libraries
library(ggplot2)
library(dplyr)
library(hrbrthemes)

combined$niche <- as.factor(combined$niche)
combined$subtype <- as.factor(combined$subtype)
```


# Figure S1F-G: Cellular subtype and pseudolocation defined gene modules and its expression across each subtype and location

```{r}
# Order clusters # 1 = 4, 2 = 6, 3 = 5, 4 = 1, 5 = 2, 6 = 3

## Cluster 1

clusternumber <- 4

cluster <- combined[which(combined$genecluster == clusternumber),] 
cluster$subtypegene <- paste(cluster$subtype, cluster$gene)
clusterAC <- cluster[which(cluster$subtype %in% c("AC")),] 
clusterMES1 <- cluster[which(cluster$subtype %in% c("MES1")),] 
clusterMES2 <- cluster[which(cluster$subtype %in% c("MES2")),] 
clusterNPC1 <- cluster[which(cluster$subtype %in% c("NPC1")),] 
clusterNPC2 <- cluster[which(cluster$subtype %in% c("NPC2")),] 
clusterOPC <- cluster[which(cluster$subtype %in% c("OPC")),] 

# "AC" = "darkolivegreen3","MES" = "tomato3", "NPC" = "darkgoldenrod1", "OPC" = "royalblue3" 
library(ggplot2)

cluster_1 <- ggplot(cluster, aes(x=niche, y=expression)) +
  geom_line(aes(group=subtypegene, color = subtype), size=0.3, alpha=0.25, show.legend = FALSE) +
  scale_colour_manual(values=c(AC = "darkolivegreen3",MES1 = "pink",MES2 = "tomato3" ,NPC1 = "darkgoldenrod1", NPC2 = "darkgoldenrod3",OPC = "royalblue3" )) +
  stat_summary(data = clusterAC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color="darkolivegreen3") +
  stat_summary(data = clusterMES1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='pink') + 
  stat_summary(data = clusterMES2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='tomato3') + 
  stat_summary(data = clusterNPC1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod1') + 
  stat_summary(data = clusterNPC2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod3') + 
  stat_summary(data = clusterOPC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='royalblue3') +theme_classic2()

grouped <- split
grouped$gene <- rownames(grouped)
genelist <- grouped[which(grouped$cutree.hclust.dist.df_norm....k...6. == clusternumber),] 
gp_up = gost(genelist$gene, organism = "mmusculus")

check_cluster1 <- gp_up$result
gostplot(gp_up, interactive = T)

## Cluster 2

clusternumber <- 6

cluster <- combined[which(combined$genecluster == clusternumber),] 
cluster$subtypegene <- paste(cluster$subtype, cluster$gene)
clusterAC <- cluster[which(cluster$subtype %in% c("AC")),] 
clusterMES1 <- cluster[which(cluster$subtype %in% c("MES1")),] 
clusterMES2 <- cluster[which(cluster$subtype %in% c("MES2")),] 
clusterNPC1 <- cluster[which(cluster$subtype %in% c("NPC1")),] 
clusterNPC2 <- cluster[which(cluster$subtype %in% c("NPC2")),] 
clusterOPC <- cluster[which(cluster$subtype %in% c("OPC")),] 

# "AC" = "darkolivegreen3","MES" = "tomato3", "NPC" = "darkgoldenrod1", "OPC" = "royalblue3" 
library(ggplot2)

cluster_2 <- ggplot(cluster, aes(x=niche, y=expression)) +
  geom_line(aes(group=subtypegene, color = subtype), size=0.3, alpha=0.25, show.legend = FALSE) +
  scale_colour_manual(values=c(AC = "darkolivegreen3",MES1 = "pink",MES2 = "tomato3" ,NPC1 = "darkgoldenrod1", NPC2 = "darkgoldenrod3",OPC = "royalblue3" )) +
  stat_summary(data = clusterAC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color="darkolivegreen3") +
  stat_summary(data = clusterMES1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='pink') + 
  stat_summary(data = clusterMES2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='tomato3') + 
  stat_summary(data = clusterNPC1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod1') + 
  stat_summary(data = clusterNPC2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod3') + 
  stat_summary(data = clusterOPC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='royalblue3') +theme_classic2()

grouped <- split
grouped$gene <- rownames(grouped)
genelist <- grouped[which(grouped$cutree.hclust.dist.df_norm....k...6. == clusternumber),] 
gp_up = gost(genelist$gene, organism = "mmusculus")

check_cluster2 <- gp_up$result
gostplot(gp_up, interactive = T)

## Cluster 3

clusternumber <- 5

cluster <- combined[which(combined$genecluster == clusternumber),] 
cluster$subtypegene <- paste(cluster$subtype, cluster$gene)
clusterAC <- cluster[which(cluster$subtype %in% c("AC")),] 
clusterMES1 <- cluster[which(cluster$subtype %in% c("MES1")),] 
clusterMES2 <- cluster[which(cluster$subtype %in% c("MES2")),] 
clusterNPC1 <- cluster[which(cluster$subtype %in% c("NPC1")),] 
clusterNPC2 <- cluster[which(cluster$subtype %in% c("NPC2")),] 
clusterOPC <- cluster[which(cluster$subtype %in% c("OPC")),] 

# "AC" = "darkolivegreen3","MES" = "tomato3", "NPC" = "darkgoldenrod1", "OPC" = "royalblue3" 
library(ggplot2)

cluster_3 <- ggplot(cluster, aes(x=niche, y=expression)) +
  geom_line(aes(group=subtypegene, color = subtype), size=0.3, alpha=0.25, show.legend = FALSE) +
  scale_colour_manual(values=c(AC = "darkolivegreen3",MES1 = "pink",MES2 = "tomato3" ,NPC1 = "darkgoldenrod1", NPC2 = "darkgoldenrod3",OPC = "royalblue3" )) +
  stat_summary(data = clusterAC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color="darkolivegreen3") +
  stat_summary(data = clusterMES1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='pink') + 
  stat_summary(data = clusterMES2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='tomato3') + 
  stat_summary(data = clusterNPC1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod1') + 
  stat_summary(data = clusterNPC2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod3') + 
  stat_summary(data = clusterOPC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='royalblue3') +theme_classic2()

grouped <- split
grouped$gene <- rownames(grouped)
genelist <- grouped[which(grouped$cutree.hclust.dist.df_norm....k...6. == clusternumber),] 
gp_up = gost(genelist$gene, organism = "mmusculus")

check_cluster3 <- gp_up$result
gostplot(gp_up, interactive = T)

## Cluster 4

clusternumber <- 1

cluster <- combined[which(combined$genecluster == clusternumber),] 
cluster$subtypegene <- paste(cluster$subtype, cluster$gene)
clusterAC <- cluster[which(cluster$subtype %in% c("AC")),] 
clusterMES1 <- cluster[which(cluster$subtype %in% c("MES1")),] 
clusterMES2 <- cluster[which(cluster$subtype %in% c("MES2")),] 
clusterNPC1 <- cluster[which(cluster$subtype %in% c("NPC1")),] 
clusterNPC2 <- cluster[which(cluster$subtype %in% c("NPC2")),] 
clusterOPC <- cluster[which(cluster$subtype %in% c("OPC")),] 

# "AC" = "darkolivegreen3","MES" = "tomato3", "NPC" = "darkgoldenrod1", "OPC" = "royalblue3" 
library(ggplot2)

cluster_4 <- ggplot(cluster, aes(x=niche, y=expression)) +
  geom_line(aes(group=subtypegene, color = subtype), size=0.3, alpha=0.25, show.legend = FALSE) +
  scale_colour_manual(values=c(AC = "darkolivegreen3",MES1 = "pink",MES2 = "tomato3" ,NPC1 = "darkgoldenrod1", NPC2 = "darkgoldenrod3",OPC = "royalblue3" )) +
  stat_summary(data = clusterAC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color="darkolivegreen3") +
  stat_summary(data = clusterMES1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='pink') + 
  stat_summary(data = clusterMES2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='tomato3') + 
  stat_summary(data = clusterNPC1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod1') + 
  stat_summary(data = clusterNPC2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod3') + 
  stat_summary(data = clusterOPC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='royalblue3') +theme_classic2()

grouped <- split
grouped$gene <- rownames(grouped)
genelist <- grouped[which(grouped$cutree.hclust.dist.df_norm....k...6. == clusternumber),] 
gp_up = gost(genelist$gene, organism = "mmusculus")

check_cluster4 <- gp_up$result
gostplot(gp_up, interactive = T)

## Cluster 5

clusternumber <- 2

cluster <- combined[which(combined$genecluster == clusternumber),] 
cluster$subtypegene <- paste(cluster$subtype, cluster$gene)
clusterAC <- cluster[which(cluster$subtype %in% c("AC")),] 
clusterMES1 <- cluster[which(cluster$subtype %in% c("MES1")),] 
clusterMES2 <- cluster[which(cluster$subtype %in% c("MES2")),] 
clusterNPC1 <- cluster[which(cluster$subtype %in% c("NPC1")),] 
clusterNPC2 <- cluster[which(cluster$subtype %in% c("NPC2")),] 
clusterOPC <- cluster[which(cluster$subtype %in% c("OPC")),] 

# "AC" = "darkolivegreen3","MES" = "tomato3", "NPC" = "darkgoldenrod1", "OPC" = "royalblue3" 
library(ggplot2)

cluster_5 <- ggplot(cluster, aes(x=niche, y=expression)) +
  geom_line(aes(group=subtypegene, color = subtype), size=0.3, alpha=0.25, show.legend = FALSE) +
  scale_colour_manual(values=c(AC = "darkolivegreen3",MES1 = "pink",MES2 = "tomato3" ,NPC1 = "darkgoldenrod1", NPC2 = "darkgoldenrod3",OPC = "royalblue3" )) +
  stat_summary(data = clusterAC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color="darkolivegreen3") +
  stat_summary(data = clusterMES1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='pink') + 
  stat_summary(data = clusterMES2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='tomato3') + 
  stat_summary(data = clusterNPC1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod1') + 
  stat_summary(data = clusterNPC2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod3') + 
  stat_summary(data = clusterOPC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='royalblue3') +theme_classic2()
grouped <- split
grouped$gene <- rownames(grouped)
genelist <- grouped[which(grouped$cutree.hclust.dist.df_norm....k...6. == clusternumber),] 
gp_up = gost(genelist$gene, organism = "mmusculus")

check_cluster5 <- gp_up$result
gostplot(gp_up, interactive = T)

## Cluster 6

clusternumber <- 3

cluster <- combined[which(combined$genecluster == clusternumber),] 
cluster$subtypegene <- paste(cluster$subtype, cluster$gene)
clusterAC <- cluster[which(cluster$subtype %in% c("AC")),] 
clusterMES1 <- cluster[which(cluster$subtype %in% c("MES1")),] 
clusterMES2 <- cluster[which(cluster$subtype %in% c("MES2")),] 
clusterNPC1 <- cluster[which(cluster$subtype %in% c("NPC1")),] 
clusterNPC2 <- cluster[which(cluster$subtype %in% c("NPC2")),] 
clusterOPC <- cluster[which(cluster$subtype %in% c("OPC")),] 

# "AC" = "darkolivegreen3","MES" = "tomato3", "NPC" = "darkgoldenrod1", "OPC" = "royalblue3" 
library(ggplot2)

cluster_6 <- ggplot(cluster, aes(x=niche, y=expression)) +
  geom_line(aes(group=subtypegene, color = subtype), size=0.3, alpha=0.25, show.legend = FALSE) +
  scale_colour_manual(values=c(AC = "darkolivegreen3",MES1 = "pink",MES2 = "tomato3" ,NPC1 = "darkgoldenrod1", NPC2 = "darkgoldenrod3",OPC = "royalblue3" )) +
  stat_summary(data = clusterAC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color="darkolivegreen3") +
  stat_summary(data = clusterMES1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='pink') + 
  stat_summary(data = clusterMES2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='tomato3') + 
  stat_summary(data = clusterNPC1, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod1') + 
  stat_summary(data = clusterNPC2, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='darkgoldenrod3') + 
  stat_summary(data = clusterOPC, aes( x=as.numeric(niche)), fun=median, geom='line',
               size=3, color='royalblue3') +theme_classic2()

grouped <- split
grouped$gene <- rownames(grouped)
genelist <- grouped[which(grouped$cutree.hclust.dist.df_norm....k...6. == clusternumber),] 
gp_up = gost(genelist$gene, organism = "mmusculus")

check_cluster6 <- gp_up$result
gostplot(gp_up, interactive = T)

# Figure S1F
ggarrange(cluster_1, cluster_2,cluster_3, cluster_4,cluster_5, cluster_6,
          ncol = 1, nrow = 6) ## Save in following scale PDF: 3.5x15

# Figure S1G
list_of_datasets <- list("Cluster1" = check_cluster1, "Cluster2" = check_cluster2, "Cluster3" = check_cluster3, "Cluster4" = check_cluster4, "Cluster5" = check_cluster5, "Cluster6" = check_cluster6)
write.xlsx(list_of_datasets, file = "~/Desktop/heatmapGO.xlsx")
```
