---
title: "Kloosterman and Erbani et al., Cell 2024: Figure 3C"
author: "Daan J. Kloosterman"
date: "21/03/24"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Figure 3C - Validation that GPNMBhigh macrophages are lipid-laden macrophages (LLMs) by sorting LLMs and looking at transcriptional overlap

# Loading packages/libraries required for the upcoming codes

```{r}
library(Seurat)
library(readxl)
library(dplyr)
library(ggplot2)
library(viridis)
library(ggpubr)
```

## Loading data

```{r}
# Set working directory to folder "pre-processed human scrna-seq" 
work_dir <- "~/surfdrive/Code Availability/Kloosterman_et_al/Kloosterman et al. 2024 - data availibility/GEO Data - processed/scrna-seq"
setwd(work_dir)

load("gbm-complete-myeloid_polished.Rda")


# Rename identity to get larger violin plots
Idents(gbm.myeloid) <- gbm.myeloid$clusterf
new.cluster.ids <- c("MG1-P2RY12", 	
                     "MG2-TNF",	
                     "MG3-GPNMB",
                     "MG4-MKI67",
                     "MDM1-CCR7",
                     "MDM2-H2-EB1",
                     "MDM3-GPNMB",
                     "MDM4-MKI67"
                     )
names(new.cluster.ids) <- levels(gbm.myeloid)
gbm.myeloid <- RenameIdents(gbm.myeloid, new.cluster.ids)
gbm.myeloid$cluster_short <- Idents(gbm.myeloid)
```

# Generating signatures based on DEG analysis on sorted (non-) LLMs from MG and MDM origin (followed standard bulk RNAseq protocol to acquire DEG list)

```{r}
## Quantifying LLMs - REAL - 
DownSig <- read.delim("~/surfdrive/Code Availability/Kloosterman_et_al/Kloosterman et al. 2024 - code availibility/signatures/DEG_LLMvsControl_Downregulated_signatures.txt")

MGspeNONLLM <- DownSig$MGspeNONLLM
MGspeNONLLM <- MGspeNONLLM[MGspeNONLLM %in% rownames(gbm.myeloid)]

MDMspeNONLLM <- DownSig$MDMspeNONLLM
MDMspeNONLLM <- MDMspeNONLLM[MDMspeNONLLM %in% rownames(gbm.myeloid)]

sharedNONLLM <- DownSig$sharedNONLLM
sharedNONLLM <- sharedNONLLM[sharedNONLLM %in% rownames(gbm.myeloid)]

UpSig <- read.delim("~/surfdrive/Code Availability/Kloosterman_et_al/Kloosterman et al. 2024 - code availibility/signatures/DEG_LLMvsControl_Upregulated_signatures.txt")

MGspeLLM <- UpSig$MGspeLLM
MGspeLLM <- MGspeLLM[MGspeLLM %in% rownames(gbm.myeloid)]

MDMspeLLM <- UpSig$MDMspeLLM
MDMspeLLM <- MDMspeLLM[MDMspeLLM %in% rownames(gbm.myeloid)]

sharedLLM <- UpSig$sharedLLM
sharedLLM <- sharedLLM[sharedNONLLM %in% rownames(gbm.myeloid)]

Gene <- list(MGspeNONLLM, MDMspeNONLLM, sharedNONLLM, MGspeLLM, MDMspeLLM, sharedLLM)

fdscores <- AddModuleScore(gbm.myeloid, features=Gene, name="Gene",nbin=100)

groups <- c("Gene1", "Gene2", "Gene3", "Gene4", "Gene5", "Gene6")

##load function to create density values
densMode <- function(x){
  td <- density(x)
  tdx <- td$x
  tdy <- td$y
  minx <- tdx[which(diff(sign(diff(tdy)))==2)]
  peakx <- tdx[which(diff(sign(diff(tdy)))==-2)]
  return(list(minx=minx, maxy=peakx))
}

##create UMAPplots
Idents(gbm.myeloid) <- gbm.myeloid$clusterf

##For every group determine the thesholds and plot several plots
for (i in groups){
  ##create densityplots and set cut-offs
  vl <- densMode(fdscores@meta.data[,i])[1]
  vl2 <- densMode(fdscores@meta.data[,i])[2]
  plot(density(fdscores@meta.data[,i]), main=paste("LLM Signature Enrichment",i, sep=""))
  if(density(fdscores@meta.data[,i])$y[which(density(fdscores@meta.data[,i])$x==vl$minx[1])] > 6 || density(fdscores@meta.data[,i])$y[which(density(fdscores@meta.data[,i])$x==vl$minx[1])] <0.01){
    abline(v=vl$minx[2])
    threshold=vl$minx[2]
  }else{
    abline(v=vl$minx[1], col="red")
    threshold=vl$minx[1]
  }
  #plot(hist(fdscores@meta.data[,i], plot=F, freq=FALSE))
  #abline(h=500)
  #abline(v=vl2$maxy, col="red")
  
  ##classify the cells based on thresholds
  gbm.myeloid@meta.data[,paste("assignedto",i, sep="")] <- "nonclassified"
  gbm.myeloid@meta.data[which(fdscores@meta.data[,i]>0.2),paste("assignedto",i, sep="")] <- paste("assignedto",i, sep=" ") ## Use 0.4 for MG
  
  ##plot figures
  plot(DimPlot(gbm.myeloid, pt.size=1,group.by=paste("assignedto",i, sep=""))+labs(title = paste("assignment based on cutoff densityplot ",i, sep=" ")))
  plot(FeaturePlot(fdscores,pt.size=1, features=i))
  plot(VlnPlot(fdscores, features=i, group.by="clusterf", pt.size = 0.0))
  plot(VlnPlot(fdscores, features=i, group.by= c("samplename"), pt.size = 0))
  DimPlot(fdscores, reduction = "umap", group.by = c("clusterf"), label = TRUE, pt.size = 0.5) 
}

sharedLLM
```

# Generate DimPlot of clusters

```{r}
a1 <- DimPlot(gbm.myeloid, reduction = "umap", group.by = "clusterf",label = F, pt.size = 0.5) 
```


# Generate FeaturePlots

```{r}
# Gene1 = MGspeNONLLM, Gene2 = MDMspeNONLLM, Gene3 = sharedNONLLM, Gene4 = MGspeLLM, Gene5 = MDMspeLLM, Gene6 = sharedLLM

a <- FeaturePlot(fdscores,pt.size=1, features="Gene1") + 
  scale_color_viridis(discrete = F) + 
  theme_bw() +
  ggtitle("MG non-LLM signature")

b <- FeaturePlot(fdscores,pt.size=1, features="Gene2") + 
  scale_color_viridis(discrete = F) + 
  theme_bw() +
  ggtitle("MDM non-LLM signature")

c <- FeaturePlot(fdscores,pt.size=1, features="Gene3") + 
  scale_color_viridis(discrete = F) + 
  theme_bw() +
  ggtitle("Common non-LLM signature")

d <- FeaturePlot(fdscores,pt.size=1, features="Gene4") + 
  scale_color_viridis(discrete = F) + 
  theme_bw() +
  ggtitle("MG LLM signature")

e <- FeaturePlot(fdscores,pt.size=1, features="Gene5") + 
  scale_color_viridis(discrete = F) + 
  theme_bw() +
  ggtitle("MDM LLM signature")

f <- FeaturePlot(fdscores,pt.size=1, features="Gene6") + 
  scale_color_viridis(discrete = F) + 
  theme_bw() +
  ggtitle("Common LLM signature")
```

# Generate VlnPlots and heatmaps of signature expression

```{r}
# Gene1 = MGspeNONLLM, Gene2 = MDMspeNONLLM, Gene3 = sharedNONLLM, Gene4 = MGspeLLM, Gene5 = MDMspeLLM, Gene6 = sharedLLM

a2 <- VlnPlot(fdscores, features="Gene1", group.by="cluster_short", pt.size = 0.0) + 
  ggtitle("MG non-LLM signature") + NoLegend()

b2 <- VlnPlot(fdscores, features="Gene2", group.by="cluster_short", pt.size = 0.0) + 
  ggtitle("MDM non-LLM signature") + NoLegend()

c2 <- VlnPlot(fdscores, features="Gene3", group.by="cluster_short", pt.size = 0.0) + 
  ggtitle("Common non-LLM signature") + NoLegend()

d2 <- VlnPlot(fdscores, features="Gene4", group.by="cluster_short", pt.size = 0.0) + 
  ggtitle("MG LLM signature") + NoLegend()

e2 <- VlnPlot(fdscores, features="Gene5", group.by="cluster_short", pt.size = 0.0) + 
  ggtitle("MDM LLM signature") + NoLegend()

f2 <- VlnPlot(fdscores, features="Gene6", group.by="cluster_short", pt.size = 0.0) + 
  ggtitle("Common LLM signature") + NoLegend()

# With labels
ggarrange(a1, 
ggarrange(a, b,c, d, e, f, 
          labels = c("B", "", "", "", "", ""),
          ncol = 3, nrow = 2),
ggarrange(a2, b2,c2, d2, e2, f2, 
          labels = c("C", "", "", "", "", ""),
          ncol = 3, nrow = 2), labels = c("A"), nrow = 3)

# Without labels (save as 12 x 17 inch portrait)
ggarrange(a1, 
ggarrange(a, b,c, d, e, f, 
          ncol = 3, nrow = 2),
ggarrange(a2, b2,c2, d2, e2, f2, 
          ncol = 3, nrow = 2), nrow = 3)

z1 <- DimPlot(gbm.myeloid, reduction = "umap", group.by = "LLM",label = F, pt.size = 0.5) + NoLegend() 
z2 <- VlnPlot(fdscores, features="Gene3", group.by="LLM", pt.size = 0.0) + 
  ggtitle("Common non-LLM signature") + NoLegend()
z3 <- VlnPlot(fdscores, features="Gene6", group.by="LLM", pt.size = 0.0) + 
  ggtitle("Common LLM signature") + NoLegend()
z1 + z2 + z3

f2
```

# Figure 3C - Violin plot of signature (from sorted LLMs and non-LLMs) enrichment in sc-RNAseq macrophage clusters

```{r}
library(corrplot)
library("PerformanceAnalytics")
library(RColorBrewer)
library(cowplot)

ident <- as.data.frame(gbm.myeloid@active.ident) 
colnames(ident) <- c("Idents")

## Adjust colors
# Subset data.frame
features =  c("non-LLM [MG]", "non-LLM [MDM]", "non-LLM [Common]", "LLM [MG]", "LLM [MDM]", "LLM [Common]")
fdscores_2 <- fdscores@meta.data
gbm.myeloid_subset <- data.frame(fdscores_2$Gene1,fdscores_2$Gene2, fdscores_2$Gene3, fdscores_2$Gene4, fdscores_2$Gene5, fdscores_2$Gene6)
colnames(gbm.myeloid_subset) <- c("non-LLM [MG]", "non-LLM [MDM]", "non-LLM [Common]", "LLM [MG]", "LLM [MDM]", "LLM [Common]")

# Add cell ID and identity classes
gbm.myeloid_subset$Cell <- rownames(gbm.myeloid_subset)
gbm.myeloid_subset <- cbind(gbm.myeloid_subset, ident)
gbm.myeloid_subset <- reshape2::melt(gbm.myeloid_subset, id.vars = c("Cell","Idents"), measure.vars = features,
                       variable.name = "Feat", value.name = "Expr")

# Figure 2C - Violin plot of signature (from sorted LLMs and non-LLMs) enrichment in sc-RNAseq macrophage clusters
ggplot(gbm.myeloid_subset, aes(Expr, factor(Idents), fill = factor(Idents))) +
        geom_violin(scale = "width", adjust = 1, trim = TRUE) +
        scale_x_continuous(expand = c(0, 0), labels = function(x)
                           c(rep(x = "", times = length(x)-2), x[length(x) - 1], "")) +
        facet_grid(cols = vars(Feat), scales = "free")  +
        theme_cowplot(font_size = 12) +
        theme(legend.position = "none", panel.spacing = unit(0, "lines"),
              panel.background = element_rect(fill = NA, color = "black"),
              strip.background = element_blank(),
              strip.text = element_text(),
              strip.text.x.top = element_text(angle = 0, hjust = 0, vjust = 0.5)) +
        ggtitle("") + xlab("Gene Expression") + ylab("")+
    scale_y_discrete(limits=rev)
```
